util/mon: reduce monitor-related allocations

mgartner opened the pull request:
#### util/mon: rename fields of MonitorName

Release note: None

#### util/uuid: add comment to uuid.Short

Release note: None

#### util/mon: add fields to MonitorName

The `processorID`, `suffix`, and `i` fields will be used in future
commits to reduce allocations related to MonitorNames.

Release note: None

#### util/mon: compile-time checks for size of BytesMonitor and BoundAccount

Release note: None

#### sql: use mon.MonitorName in more places

Using structured `mon.MonitorName` objects reduces allocations of
strings in some cases. This commit doesn't replace all usages of string
monitor names - there are myriad such usages. We can iteratively chip
away at them in future commits.

Release note: None

#### util/mon: reduce size of MonitorName

The size of `MonitorName` has been reduced by replacing the string
suffix with a suffix enum.

Release note: None

#### util/mon: further reduce the size of MonitorName

The size of `MonitorName` has been reduced by using the same 4 bytes to
store either the `int32` processor ID or the `uuid.Short`, and by
reducing the optional integer suffix from an `int32` to `uint16`, which
should be sufficient in all cases.

Release note: None

#### util/mon: rename MonitorName to Name

Release note: None

#### util/mon: remove deprecated NewMonitorWithStringName

All usages of the deprecated NewMonitorWithStringName function have been
replaced with NewMonitor, and the former has been removed.

Release note: None

#### util/mon: remove deprecated NewMonitorInheritWithLimitAndStringName

All usages of the deprecated NewMonitorInheritWithLimitAndStringName
function have been replaced with NewMonitorInheritWithLimit, and the
former has been removed.

Release note: None

#### colexec: batch allocations of BoundAccount

Allocations of `BoundAccount`s in `createUnlimitedMemAccountsLocked` are
now batched into a single allocation.

Release note: None

#### sql/rowflow: refactor monitor name

The monitor names created in `(*rowBasedFlow).setupRouter` now use
`(*mon.Name).WithID` to add the stream ID to the name instead of
allocating a new string.

Epic: None
Release note: None

mgartner committed:
util/mon: rename fields of MonitorName

Release note: None

mgartner committed:
util/uuid: add comment to uuid.Short

Release note: None

mgartner committed:
util/mon: add fields to MonitorName

The `processorID`, `suffix`, and `i` fields will be used in future
commits to reduce allocations related to MonitorNames.

Release note: None

mgartner committed:
util/mon: compile-time checks for size of BytesMonitor and BoundAccount

Release note: None

mgartner committed:
sql: use mon.MonitorName in more places

Using structured `mon.MonitorName` objects reduces allocations of
strings in some cases. This commit doesn't replace all usages of string
monitor names - there are myriad such usages. We can iteratively chip
away at them in future commits.

Release note: None

mgartner committed:
util/mon: reduce size of MonitorName

The size of `MonitorName` has been reduced by replacing the string
suffix with a suffix enum.

Release note: None

mgartner committed:
util/mon: further reduce the size of MonitorName

The size of `MonitorName` has been reduced by using the same 4 bytes to
store either the `int32` processor ID or the `uuid.Short`, and by
reducing the optional integer suffix from an `int32` to `uint16`, which
should be sufficient in all cases.

Release note: None

mgartner committed:
util/mon: rename MonitorName to Name

Release note: None

mgartner committed:
colexec: batch allocations of BoundAccount

Allocations of `BoundAccount`s in `createUnlimitedMemAccountsLocked` are
now batched into a single allocation.

Release note: None

mgartner committed:
util/mon: remove deprecated NewMonitorWithStringName

All usages of the deprecated NewMonitorWithStringName function have been
replaced with NewMonitor, and the former has been removed.

Release note: None

mgartner committed:
util/mon: remove deprecated NewMonitorInheritWithLimitAndStringName

All usages of the deprecated NewMonitorInheritWithLimitAndStringName
function have been replaced with NewMonitorInheritWithLimit, and the
former has been removed.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143305)
<!-- Reviewable:end -->

mgartner commented:
It's a very minor reduction in allocations. But it is nice to have structured names instead of strings.

```
name                                              old allocs/op  new allocs/op  delta
FlowSetup/vectorize=true/distribute=true               162 ± 0%       161 ± 3%  -0.80%  (p=0.000 n=16+17)
FlowSetup/vectorize=true/distribute=false              137 ± 0%       136 ± 5%  -0.86%  (p=0.000 n=16+17)
FlowSetup/vectorize=false/distribute=true              160 ± 0%       160 ± 0%    ~     (all equal)
FlowSetup/vectorize=false/distribute=false             135 ± 0%       135 ± 0%    ~     (all equal)
```

yuzefovich reviewed (approved):
Nice cleanup! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, 33 of 33 files at r5, 8 of 8 files at r6, 3 of 3 files at r7, 77 of 77 files at r8, 17 of 17 files at r9, 14 of 14 files at r10, 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143305#-:-OM8Z_mL4OgmW8r-sczM:b-jy7zf3)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @asg0451, @jbowens, @kev-cao, and @xinhaoz)

___
*[`pkg/util/mon/bytes_usage.go` line 281 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/143305#-OM8X_iI1mMlYKW9x6Fb:-OM8X_iI1mMlYKW9x6Fc:bjrymme) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe874906e8e950138cfc54213cbb29ed439d4793/pkg/util/mon/bytes_usage.go#L281)):*
> ```Go
> }
> 
> // Name is used to identify monitors in logging messages. It consists of:
> ```

nit: let's also mention `crdb_internal.node_memory_monitors` virtual table where the monitor name shows up.
___
*[`pkg/sql/rowflow/row_based_flow.go` line 464 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/143305#-OM8YzK72fb4O5If102_:-OM8YzK72fb4O5If102a:b-q95jpz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8d6d445b7a98d0e9ba074c42c849c4e1f5036db/pkg/sql/rowflow/row_based_flow.go#L464)):*
> ```Go
> 	diskMonitors := make([]*mon.BytesMonitor, len(spec.Streams))
> 	for i := range spec.Streams {
> 		// TODO(mgartner): Is a StreamID ever greater than math.MaxInt32? If
> ```

I don't think it can ever exceed max int32 in practice. `StreamID`s are indexes into `PhysicalInfrastructure.Streams` slice, which contains all pair-wise "links" between two processors within the same plan, so I'd expect that we'd OOM-crash if stream ID reaches 1 million, way before max int 32.


<!-- Sent from Reviewable.io -->

mgartner committed:
sql/rowflow: refactor monitor name

The monitor names created in `(*rowBasedFlow).setupRouter` now use
`(*mon.Name).WithID` to add the stream ID to the name instead of
allocating a new string.

Release note: None

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143305)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asg0451, @jbowens, @kev-cao, @xinhaoz, and @yuzefovich)

___
*[`pkg/sql/rowflow/row_based_flow.go` line 464 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/143305#-OM8YzK72fb4O5If102_:-OME9B9q6uA_p7wsxgYX:bqj0ob1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8d6d445b7a98d0e9ba074c42c849c4e1f5036db/pkg/sql/rowflow/row_based_flow.go#L464)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I don't think it can ever exceed max int32 in practice. `StreamID`s are indexes into `PhysicalInfrastructure.Streams` slice, which contains all pair-wise "links" between two processors within the same plan, so I'd expect that we'd OOM-crash if stream ID reaches 1 million, way before max int 32.
</blockquote></details>

Makes sense, I'll add another commit to address this TODO then. Thanks!
___
*[`pkg/util/mon/bytes_usage.go` line 281 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/143305#-OM8X_iI1mMlYKW9x6Fb:-OMEAVhs99Z0UujO8k7c:b-rnraql) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe874906e8e950138cfc54213cbb29ed439d4793/pkg/util/mon/bytes_usage.go#L281)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: let's also mention `crdb_internal.node_memory_monitors` virtual table where the monitor name shows up.
</blockquote></details>

done.


<!-- Sent from Reviewable.io -->

mgartner commented:
TFTR!

bors r+
